sled-agent/src/instance_manager.rs

```diff
@@ -90,43 +92,66 @@ impl InstanceManager {
 
         let (instance, maybe_instance_ticket) = {
             let mut instances = self.inner.instances.lock().unwrap();
-            if let Some(instance) = instances.get_mut(&instance_id) {
-                // Instance already exists.
-                info!(&self.inner.log, "instance already exists");
-                (instance.clone(), None)
-            } else {
-                // Instance does not exist - create it.
-                info!(&self.inner.log, "new instance");
-                let instance_log = self
-                    .inner
-                    .log
-                    .new(o!("instance" => instance_id.to_string()));
-                instances.insert(
-                    instance_id,
-                    Instance::new(
-                        instance_log,
+            match instances.get_mut(&instance_id) {
+                Some(instance) if migrate.is_none() => {
+                    // Instance already exists and we're not performing a migration
+                    info!(&self.inner.log, "instance already exists");
+                    (instance.clone(), None)
+                }
+                _ => {
+                    // Instance does not exist or one does but we're performing
+                    // a intra-sled migration. Either way - create an instance
```

Comment:
What if "ensure" has been called repeatedly for this migration? I'm trying to eyeball how this function can still be idempotent if the following behavior occurs:

1. "Ensure" called, requesting a migration. Instance now exists.
2. "Ensure" called again, while the migration is in-progress, because a saga step is being called repeatedly.

It seems like this results in creating a brand new propolis instance, and terminating the old instance, no? Couldn't this potentially interrupt migration in progress, by basically "throwing away the destination" instance and creating it again?

(Maybe this is intended behavior, but I figured I'd check - kinda seems like we'd want to leave the destination alone, or potentially just alter attached devices, if it is already communicating with the source instance)

Reply:
I guess, I'd probably also need to pass along both the src & dst UUIDs in the migrate request here so that we can determine if the migration already happened? Admittedly the idempotent story is a little ill-defined so far and is definitely something I need to audit more fine-grainedly.

Reply:
Yeah, this seems possible to infer a handful of ways:

- If the "source_propolis" UUID == "destination_propolis" UUID
- If the migration UUID was seen before

I do still think it's important to handle though, otherwise, the behavior of tossing the old instance would disrupt the migration.

Reply:
Ok, updated it such that if the current instance's propolis UUID matches the target propolis UUID, we treat it as an ongoing migration and not spin up a new instance.